File: haskell/grains/src/Grains.hs
module Grains (square, total) where

square :: Integer -> Maybe Integer
square n
  | n < 1 || 64 < n = Nothing
  | otherwise = Just $ 2 ^ (n - 1)

total :: Integer
total = sum $ map (2 ^) ([0 .. 64 - 1] :: [Integer])

File: c/grains/src/grains.c
#include "grains.h"

uint64_t square(uint8_t index)
{
    return (1 <= index && index <= 64)
               ? UINT64_C(1) << (index - 1) // 2^(index-1)
               : 0;
}

uint64_t total(void)
{
    return UINT64_MAX; // (2^64)-1
}